RHIDP-4807: Third-party plugin installation in RHDH (#733)
redhat-developer/red-hat-developers-documentation-rhdh: modules/dynamic-plugins/proc-package-third-party-dynamic-plugin.adoc

[id="proc-package-third-party-dynamic-plugin_{context}"]
= Packaging and publishing third-party plugins as dynamic plugins

To package a third-party plugin as a dynamic plugin, you need to access to its source code and must export a plugin using the `@janus-idp/cli`. For more information, see xref:proc-export-third-party-plugins-rhdh_{context}[]. 

> After exporting a third-party plugin, you can convert the package into one of the following supported formats:

Reviewer: ```suggestion
After exporting a third-party plugin, you can package the derived package into one of the following supported formats:
```

* Open Container Initiative (OCI) image (recommended)
* TGZ file
* JavaScript package

> 

Reviewer: [IMPORTANT]
====
Red Hat recommends packaging dynamic plugins into Open Container Initiative (OCI) image 
====

Author: (Recommended) is added along with the Open Container Initiative (OCI) image in the first bullet.

[IMPORTANT]
====
Exported dynamic plugin packages must only be published to private NPM registries.
====

[#convert-to-oci-image]

> == Converting a dynamic plugin package to an OCI image

Reviewer: ```suggestion
== Packaging a dynamic plugin packages to an OCI image
```
or "Creating OCI Image with dynamic packages". There is no conversion

.Prerequisites
* You have installed `podman` or `docker`.
* You have exported a third-party dynamic plugin package. For more information, see xref:proc-export-third-party-plugins-rhdh_{context}[].

.Procedure

> . Navigate to the exported plugin's root directory (not the `dist-dynamic` directory).

Reviewer: ```suggestion
. Navigate to the plugin's root directory (not the `dist-dynamic` directory).
```
it should not say "exported" as the exported plugin is on in the dist-dynamic

. Run the following command to package the plugin into an OCI image:
+
--
.Example command to package an exported third-party plugin
[source,bash]
----
npx @janus-idp/cli@latest package package-dynamic-plugins --tag quay.io/example/image:v0.0.1
----

In the previous command, the `--tag` argument specifies the image name and tag.
--
. Run one of the following commands to push the image to a registry:
+
--
.Example command to push an image to a registry using podman
[source,bash]
----
podman push quay.io/example/image:v0.0.1
----

.Example command to push an image to a registry using docker
[source,bash]
----
docker push quay.io/example/image:v0.0.1
----

The output of the previous commands provides the plugin's path for use in the `dynamic-plugin-config.yaml` file.
--

[#convert-to-tgz]

> == Converting a dynamic plugin package to a TGZ file

Reviewer: ```suggestion
== Packaging a dynamic plugin packages to a TGZ file
```

Author: Changed it to "== Creating a TGZ file with dynamic packages"

.Prerequisites
* You have exported a third-party dynamic plugin package. For more information, see xref:proc-export-third-party-plugins-rhdh_{context}[].

.Procedure
. Navigate to the `dist-dynamic` directory.
. Run the following command to create a `tgz` archive:
+
--
.Example command to create a `tgz` archive
[source,bash]
----
npm pack
----
--
. To retrieve the integrity hash of the archive, run the following command:
+
--
.Example command to retrieve the integrity hash of a `tgz` archive
[source,bash]
----

> npm pack --json | jq -r '.[0].integrity'

Reviewer: Not everyone might have `jq` installed, or we need to add it to the prerequisites.
This is why in the original document had `npm pack --json | head -n 10` as the main command for users to use.

Author: Fixed it.

----
--
. Host the archive on a web server accessible to your {product-very-short} instance, and reference its URL in the `dynamic-plugin-config.yaml` file as follows:
+
--
.Example `dynamic-plugin-config.yaml` file
[source,yaml]
----
plugins:
  - package: https://example.com/backstage-plugin-myplugin-1.0.0.tgz
    integrity: sha512-<hash>
----
--
. Run the following command to package the plugins:
+
--

> .Example command to package a dynamic plugin

Reviewer: This should be a separate flow from the main procedure. 
It should be titled  "Example commands to build and deploy an HTTP server {ocp-short}"

Author: Added the commands in a Tip block.

[source,bash]
----
npm pack --pack-destination ~/test/dynamic-plugins-root/
----
--
. Build and deploy an HTTP server in {ocp-short}:
+
--
.Example commands to build and deploy an HTTP server in {ocp-short}
[source,bash]
----
oc project rhdh
oc new-build httpd --name=plugin-registry --binary
oc start-build plugin-registry --from-dir=dynamic-plugins-root --wait
oc new-app --image-stream=plugin-registry
----
--
. Configure your {product-very-short} to use plugins from the HTTP server by editing the `dynamic-plugin-config.yaml` file:
+
--
.Example configuration to use packaged plugins in {product-very-short} 
[source,yaml]
----
plugins:
  - package: http://plugin-registry:8080/backstage-plugin-myplugin-1.9.6.tgz
----
--

[#convert-to-jsp]

> == Converting a dynamic plugin to a JavaScript package

Reviewer: ```suggestion
== Packaging dynamic plugins to a JavaScript package
```

Author: Changed it to "== Creating a JavaScript package with dynamic packages"

> 

Reviewer: The following warning is important and should be repeated here again.
```suggestion
+
WARNING: The derived dynamic plugin JavaScript packages should not be pushed to the public npm registry. If you need to push it to npm registry, use your own private one.
+
```

[id="proc-package-third-party-dynamic-plugin_{context}"]

> = Packaging and publishing third-party plugins as dynamic plugins

Reviewer 2: Consider transforming it into an assembly including multiple procedures.

> To package a third-party plugin as a dynamic plugin, you need to access to its source code and must export a plugin using the `@janus-idp/cli`. For more information, see xref:proc-export-third-party-plugins-rhdh_{context}[]. 

Reviewer 3: "...and must export a plugin using `@janus-idp/cli'."

I think "the" is not necessary before '@janus-idp/cli'...

Author: +1, removed the article

Author: This paragraph has been removed.

> To package a third-party plugin as a dynamic plugin, you need to access to its source code and must export a plugin using the `@janus-idp/cli`. For more information, see xref:proc-export-third-party-plugins-rhdh_{context}[]. 
> 
> After exporting a third-party plugin, you can package the derived package into one of the following supported formats:

Reviewer 2: Consider removing fluff and repeated content

```suggestion
After xref:proc-export-third-party-plugins-rhdh_{context}[exporting a third-party plugin], you can package the derived package into one of the following supported formats:
```

* Open Container Initiative (OCI) image (recommended)
* TGZ file
* JavaScript package

[IMPORTANT]
====
Exported dynamic plugin packages must only be published to private NPM registries.
====

[#create-oci-image]

> == Creating an OCI image with dynamic packages

Reviewer 2: Consider moving the procedure to a dedicated file.

Author: Done

--

[#create-tgz-file]
== Creating a TGZ file with dynamic packages

.Prerequisites
* You have exported a third-party dynamic plugin package. For more information, see xref:proc-export-third-party-plugins-rhdh_{context}[].

.Procedure
. Navigate to the `dist-dynamic` directory.
. Run the following command to create a `tgz` archive:
+
--
.Example command to create a `tgz` archive
[source,bash]
----
npm pack
----
You can obtain the integrity hash from the output of the `npm pack` command by using the `--json` flag as follows:

.Example command to obtain the integrity hash of a `tgz` archive
[source,bash]
----
npm pack --json | head -n 10
----
--

. Host the archive on a web server accessible to your {product-very-short} instance, and reference its URL in the `dynamic-plugin-config.yaml` file as follows:
+
--
.Example `dynamic-plugin-config.yaml` file
[source,yaml]
----
plugins:
  - package: https://example.com/backstage-plugin-myplugin-1.0.0.tgz
    integrity: sha512-<hash>
----
--
. Run the following command to package the plugins:
+
--
.Example command to package a dynamic plugin
[source,bash]
----
npm pack --pack-destination ~/test/dynamic-plugins-root/
----

> [TIP]

Reviewer: 👍  Marking this a "TIP" is perfect. I did not know that we had this in our docs  👍

====
To create a plugin registry using HTTP server on {ocp-short}, run the following commands:

.Example commands to build and deploy an HTTP server in {ocp-short}
[source,bash]
----
oc project rhdh
oc new-build httpd --name=plugin-registry --binary
oc start-build plugin-registry --from-dir=dynamic-plugins-root --wait
oc new-app --image-stream=plugin-registry
----
====
--

. Configure your {product-very-short} to use plugins from the HTTP server by editing the `dynamic-plugin-config.yaml` file:
+
--
.Example configuration to use packaged plugins in {product-very-short} 
[source,yaml]
----
plugins:
  - package: http://plugin-registry:8080/backstage-plugin-myplugin-1.9.6.tgz
----
--

[#create-js-package]

> == Creating a JavaScript package with dynamic packages

Reviewer 2: Consider moving the procedure to a dedicated file.

[WARNING]
====
The derived dynamic plugin JavaScript packages must not be published to the public NPM registry. If you must publish to the NPM registry, use a private registry.
====

.Prerequisites
* You have exported a third-party dynamic plugin package. For more information, see xref:proc-export-third-party-plugins-rhdh_{context}[].

.Procedure
. Navigate to the `dist-dynamic` directory.
. Run the following command to publish the package to your private NPM registry:
+
--
.Example command to publish a plugin package to an NPM registry
[source,bash]
----
npm publish --registry <npm_registry_url>
----

[TIP]
====
You can add the following to your `package.json` file before running the `export` command:

.Example `package.json` file
[source,json]
----
{
  "publishConfig": {
    "registry": "<npm_registry_url>"
  }
}
----

> If you modify the `publishConfig` after exporting the dynamic plugin, re-run the `export-dynamic-plugin` command to ensure the correct configuration is included.

Reviewer 3: I would suggest removing "the" before 'publishConfig' as we did not include an identifier after that.
